src/mam/mod_mam_rdbms_arch_async.erl

```diff
@@ -0,0 +1,190 @@
+-spec start(mongooseim:host_type(), [{writer_type(), gen_mod:module_opts()}]) -> any().
+start(HostType, Opts) ->
+    ?LOG_INFO(#{opts => Opts, what => starting_mam_async_pool, host_type => HostType}),
+    [ start_pool(HostType, Mod) || Mod <- Opts ].
+
+-spec stop(mongooseim:host_type()) -> any().
+stop(HostType) ->
+    ?LOG_INFO(#{what => stopping_mam_async_pool, host_type => HostType}),
+    Opts = gen_mod:get_module_opts(HostType, ?MODULE),
+    [ stop_pool(HostType, Mod) || Mod <- Opts ].
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    mongoose_async_pools:config_spec().
+
+-spec supported_features() -> [atom()].
+supported_features() ->
+    [dynamic_domains].
+
+%%%===================================================================
+%%% internal callbacks
+%%%===================================================================
+-spec start_pool(mongooseim:host_type(), {writer_type(), gen_mod:module_opts()}) ->
+    supervisor:startchild_ret().
+start_pool(HostType, {Type, Opts}) ->
+    Opts1 = extend_opts(Type, Opts),
+    prepare_insert_queries(Type, Opts1),
+    ensure_metrics(Type, HostType),
+    register_hooks(Type, HostType),
+    start_pool(Type, HostType, Opts1).
+
+extend_opts(Type, Opts) ->
+    Merge = maps:merge(defaults(), maps:from_list(Opts)),
+    Opts1 = maps:to_list(Merge),
+    Opts2 = gen_mod:set_opt(flush_extra, Opts1, Merge),
+    add_callback(Type, Opts2).
+
+add_callback(pm, Opts) ->
+    gen_mod:set_opt(flush_callback, Opts, fun ?MODULE:flush_pm/2);
+add_callback(muc, Opts) ->
+    gen_mod:set_opt(flush_callback, Opts, fun ?MODULE:flush_muc/2).
+
+defaults() ->
+    #{flush_interval => 2000,
+      batch_size => 30,
+      pool_size => 2 * erlang:system_info(schedulers_online)}.
+
+prepare_insert_queries(pm, Opts) ->
+    MaxSize = gen_mod:get_opt(batch_size, Opts, 30),
+    mod_mam_rdbms_arch:prepare_insert(insert_mam_message, 1),
+    mod_mam_rdbms_arch:prepare_insert(insert_mam_messages, MaxSize);
+prepare_insert_queries(muc, Opts) ->
+    MaxSize = gen_mod:get_opt(batch_size, Opts, 30),
+    mod_mam_muc_rdbms_arch:prepare_insert(insert_mam_muc_message, 1),
+    mod_mam_muc_rdbms_arch:prepare_insert(insert_mam_muc_messages, MaxSize).
+
+ensure_metrics(pm, HostType) ->
+    mongoose_metrics:ensure_metric(HostType, ?PM_PER_MESSAGE_FLUSH_TIME, histogram),
+    mongoose_metrics:ensure_metric(HostType, ?PM_FLUSH_TIME, histogram);
+ensure_metrics(muc, HostType) ->
+    mongoose_metrics:ensure_metric(HostType, ?MUC_PER_MESSAGE_FLUSH_TIME, histogram),
+    mongoose_metrics:ensure_metric(HostType, ?MUC_FLUSH_TIME, histogram).
+
+register_hooks(pm, HostType) ->
+    ejabberd_hooks:add(mam_archive_message, HostType, ?MODULE, archive_pm_message, 50);
+register_hooks(muc, HostType) ->
+    ejabberd_hooks:add(mam_muc_archive_message, HostType, ?MODULE, archive_muc_message, 50).
+
+start_pool(pm, HostType, Opts) ->
+    mongoose_async_pools:start_pool(HostType, 'pm-mam', Opts);
```

Comment:
maybe `mam_pm`. We almost never use `-` as a separator, because it requires extra escaping